### DISCUSSION TITLE
Cannot get overloaded constant for type ascription, when qualifing `as` a function

### DISCUSSION THREAD

**Discussion by chansey97** [cvc5/cvc5#7274] [Q&A] [answered]
**Describe the bug**

```
cvc5> (set-logic ALL)
cvc5> (declare-fun f (Int) Int)
cvc5> (declare-fun g1 (Int) Int)
cvc5> (assert (forall ((x Int))  (= ((as f Int) x) (g1 x))))
(error "Parse Error: <shell>:1.39: Type ascription not satisfied, term f expected sort Int but has sort (-> Int Int)

  (assert (forall ((x Int))  (= ((as f Int) x) (g1 x))))
                                       ^
")
cvc5>
```

According the standard http://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf, page 27:

>   Recall that every function symbol f is separately associated with one or more ranks, each specifying the sorts of f ’s arguments and result. To simplify sort checking, a function symbol in a term can be annotated with one of its result sorts σ. Such an annotated function symbol is a qualified identifier of the form (as f σ).

I think we should be able to use the result sorts σ as `f`'s sort, but cvc4/cvc5 raised an error.

PS. z3 has no this problem.

**Command line arguments**: 

```
cvc5 -m --lang smt
```

**cvc5 version/commit**: 

https://github.com/CVC4/CVC4/releases/download/1.8/cvc4-1.8-win64-opt.exe

or

http://cvc4.cs.stanford.edu/downloads/builds/win64-opt/unstable/cvc5-2021-09-29-win64-opt.exe

**Operating system**: 

Windows 7

**Comment by ajreynol** [comment, marked as answer]
I agree this is an issue, the fix should be relatively straightforward.
